ipc/provider/src/checkpoint.rs

```diff
@@ -162,30 +178,62 @@ impl<T: BottomUpCheckpointRelayer + Send + Sync + 'static> BottomUpCheckpointMan
                     .await?;
                 log::debug!("bottom up bundle: {bundle:?}");
 
-                let epoch = self
-                    .parent_handler
-                    .submit_checkpoint(
-                        submitter,
-                        bundle.checkpoint,
-                        bundle.signatures,
-                        bundle.signatories,
-                    )
+                // We support parallel checkpoint submission using FIFO order with a limited parallelism (controlled by
+                // the size of submission_semaphore).
+                // We need to acquire a permit (from a limited permit pool) before submitting a checkpoint.
+                // We may wait here until a permit is available.
+                let parent_handler_clone = Arc::clone(&self.parent_handler);
+                let submission_permit = self
+                    .submission_semaphore
+                    .clone()
+                    .acquire_owned()
                     .await
-                    .map_err(|e| {
-                        anyhow!(
-                            "cannot submit bottom up checkpoint at height {} due to: {e:}",
-                            event.height
-                        )
-                    })?;
-
-                log::info!(
-                    "submitted bottom up checkpoint({}) in parent at height {}",
-                    event.height,
-                    epoch
+                    .unwrap();
```

Comment:
Just curious about the behaviour when the number of checkpoints to submit (say 10) is actually more than number of permits (say 2). So `acquire_owned` will actually block if 2 checkpoints are being submitted until one of 2 actually finishes? Even if `submit_checkpoint ` fails and the thread crashes, `drop(submission_permit)` will be executed and no dead lock here right?

Reply:
> So acquire_owned will actually block if 2 checkpoints are being submitted until one of 2 actually finishes? 

Yes, it will literally block (the `for` loop just pauses).

> Even if submit_checkpoint  fails and the thread crashes, drop(submission_permit) will be executed and no dead lock here right?

My last version had an issue where the async task can panic before dropping the permit (because the use of `unwrap()`). It's been fixed here so the permit will be dropped no matter it succeeds or fails when submitting the checkpoint.